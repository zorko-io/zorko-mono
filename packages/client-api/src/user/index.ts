export * from './user.api.client';
export * from './user.axios.api.client';
