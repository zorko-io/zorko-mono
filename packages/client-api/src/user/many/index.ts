export * from './user.many.api.client'
export * from './user.axios.many.api.client'
