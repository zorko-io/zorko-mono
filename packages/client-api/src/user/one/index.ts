export * from './user.one.api.client'
export * from './user.axios.one.api.client'
