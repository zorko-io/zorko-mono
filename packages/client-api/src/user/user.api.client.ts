import { AuthApiClient } from '../auth';
import { RemoteUserApi } from '@zorko/remote-api';

export interface UserApiClient extends AuthApiClient, RemoteUserApi {}
