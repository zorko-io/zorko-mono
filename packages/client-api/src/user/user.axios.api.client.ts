import { AxiosResponse } from 'axios';
import { AuthAxiosApiClient } from '../auth';
import { UserDto } from '@zorko/dto';
import { UserApiClient } from './user.api.client';
import {
  CreateUserCollectionParams,
  CreateUserParams, DeleteUserCollectionParams, DeleteUserParams,
  ReadUserCollectionParams,
  ReadUserParams,
  UpdateUserCollectionParams, UpdateUserParams,
} from '@zorko/remote-api';

export class UserAxiosApiClient extends AuthAxiosApiClient implements UserApiClient  {

  async findMany(readParams?: ReadUserCollectionParams): Promise<UpdateUserCollectionParams> {
    const response: AxiosResponse<UpdateUserCollectionParams> = await this.http.get('/users');
    return response.data
  }

  async createOne(user: CreateUserParams): Promise<string> {
    const response: AxiosResponse<string> = await this.http.post('/users', user);
    return response.data;
  }

  async findOne(readParams: ReadUserParams): Promise<UpdateUserParams> {
    const { id }  = readParams;
    const response: AxiosResponse<UserDto> = await this.http.get(`/users/${id}`);
    return response.data;
  }

  async createMany(createParams: CreateUserCollectionParams): Promise<string> {
    return '123'
  }

  async removeMany(deleteParams: DeleteUserCollectionParams): Promise<string> {
    const response: AxiosResponse<string> = await this.http.delete('/users');
    return response.data
  }

  async removeOne(deleteParams: DeleteUserParams): Promise<void> {
    const { id } = deleteParams;
    await this.http.delete(`/users/${id}`);
  }

  updateMany(updateParams: UpdateUserCollectionParams): Promise<ReadUserCollectionParams> {
    return undefined;
  }

  updateOne(updateParams: UpdateUserParams): Promise<ReadUserParams> {
    return undefined;
  }
}

