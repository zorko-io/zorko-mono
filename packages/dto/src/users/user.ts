import { UserDto } from './user.dto';
import { UserDtoInterface } from './user.dto.interface';
import { RolesEnum } from '../roles';
import { validateSync, ValidationError } from 'class-validator';

export class User {

  private dto: UserDto;

  constructor(email: string) {
    this.dto = new UserDto();
    this.dto.email = email;
  }

  getId(): string {
    return this.dto.id;
  }

  setId(id: string): this {
    this.dto.id = id;
    return this;
  }

  getEmail(): string {
    return this.dto.email;
  }

  setEmail(email: string): this {
    this.dto.email = email;
    return this;
  }

  getPassword(): string  {
    return this.dto.password;
  }

  setPassword(password: string): this {
    this.dto.password = password;
    return this;
  }

  getRoles(): RolesEnum[] {
    return this.hasRoles() ? this.dto.roles : [];
  }

  hasRoles () : boolean {
    return this.dto.roles && this.dto.roles.length > 0
  }

  setRoles(roles: RolesEnum[]): this {
    this.dto.roles = roles;
    return this;
  }

  validate () : ValidationError[] {
     return validateSync(this.dto);
  }

  toDTO(): UserDtoInterface {

    const errors = this.validate();

    if (errors.length){
      throw { code: 'VALIDATION_ERROR', errors }
    }

    return {
      id: this.dto.id,
      email: this.dto.email,
      password: this.dto.password,
      roles: this.dto.roles
    }
  }
}
