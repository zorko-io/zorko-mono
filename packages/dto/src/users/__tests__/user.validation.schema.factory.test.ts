import { User } from '../user';
import { userValidationSchema } from '../user.validation.schema';

describe('userValidationSchema', () => {
  let schema;
  let data: User;
  let actual;

  beforeEach(() => {
    schema = userValidationSchema();
  });

  it('passes minimum validation', () => {
    data = { email: 'test@ffkfkf.com', password: '34324' };
    actual = schema.isValidSync(data);

    expect(actual).toBeTruthy();
  });

  it('passes max validation', () => {
    data = {
      id: '312123',
      email: 'test@ffkfkf.com',
      roles: [],
      password: '234342423434'
    };
    actual = schema.isValidSync(data);

    expect(actual).toBeTruthy();
  });

  it('passes `hashPassword` instead of real password', () => {
    data = {
      id: '312123',
      email: 'test@ffkfkf.com',
      roles: [],
      hashPassword: 'nfcjnd93024984930840'
    };

    actual = schema.isValidSync(data);

    schema.validateSync(data);


    expect(actual).toBeTruthy();
  });

  it('fails on email', () => {
    let data = {
      id: '312123',
      email: 'test!ffkfkf_94949494',
      roles: [],
      password: '234342423434'
    };
    actual = schema.isValidSync(data);

    expect(actual).toBeFalsy();

    try {
      schema.validateSync(data);
    } catch (error) {
      expect(error.path).toMatchSnapshot();
      expect(error.message).toMatchSnapshot();
    }
  });

  it('fails on all invalid', () => {
    let data = {
      id: 'fsdfd',
      email: 'test!ffkfkf_94949494',
      roles: {},
      password: {}
    };
    actual = schema.isValidSync(data);

    expect(actual).toBeFalsy();

    try {
      schema.validateSync(data, { abortEarly: false });
    } catch (error) {
      expect(error.inner).toMatchSnapshot();
    }
  });

});
