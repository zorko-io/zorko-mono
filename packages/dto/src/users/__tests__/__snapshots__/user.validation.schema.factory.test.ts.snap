// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserValidationSchemaFactory fails on all invalid 1`] = `
Array [
  [ValidationError: email must be a valid email],
  [ValidationError: roles must be a \`array\` type, but the final value was: \`null\` (cast from the value \`{}\`).
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`],
]
`;

exports[`UserValidationSchemaFactory fails on email 1`] = `"email"`;

exports[`UserValidationSchemaFactory fails on email 2`] = `"email must be a valid email"`;

exports[`userValidationSchema fails on all invalid 1`] = `
Array [
  [ValidationError: email must be a valid email],
  [ValidationError: roles must be a \`array\` type, but the final value was: \`null\` (cast from the value \`{}\`).
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`],
]
`;

exports[`userValidationSchema fails on email 1`] = `"email"`;

exports[`userValidationSchema fails on email 2`] = `"email must be a valid email"`;
